yt/utilities/lib/cyoctree.pyx

```diff
@@ -1,679 +1,756 @@
-             self._data_version) = \
-                struct.unpack('2Q3iq', f.read(40))
-            self._idx = \
-                np.asarray(struct.unpack('{}Q'.format(self.num_particles),
-                           f.read(8*self.num_particles)), dtype=np.int64)
-
-            reserve(&self.nodes, num_octs+1)
-            for i in range(num_octs):
-                (temp.left_edge[0], temp.left_edge[1], temp.left_edge[2],
-                 temp.right_edge[0], temp.right_edge[1], temp.right_edge[2],
-                 temp.start, temp.end, temp.parent, temp.children, temp.leaf,
-                 temp.node_id, temp.leaf_id, temp.depth) = \
-                struct.unpack('6d4Q?2QB', f.read(105))
-                self.nodes.push_back(temp)
-
-        for i in range(3):
-            self._left_edge[i] = self.nodes[0].left_edge[i]
-            self._right_edge[i] = self.nodes[0].right_edge[i]
+                                    hsml[i], prefactor, prefactor_norm,
+                                    0, use_normalization=use_normalization)
+
 
 @cython.boundscheck(False)
 @cython.wraparound(False)
 @cython.cdivision(True)
-cdef int split_helper(int start, int end, int num_children, int max_splits,
-                       np.int64_t * split_arr, np.float64_t * pos,
-                       np.int64_t * idx, np.float64_t * left,
-                       np.float64_t * right) nogil except -1:
-    '''
-    This takes in the split array and sets up the first and last particle, it
-    then calls the spit function.
-    '''
-    cdef double lower = left[0]
-    cdef double upper = right[0]
-
-    split_arr[0] = start
-    split_arr[num_children] = end
-
-    split(0, num_children, max_splits, 0, 0, split_arr, idx, pos, left, right,
-          lower, upper)
+cdef int seperate_x(Octree * octree, double value, np.int64_t start, np.int64_t end) nogil except -1:
+    """
+    This is a simple utility function which takes a selection of particles and
+    re-arranges them such that values below `value` are to the left of split and
+    values above are to the right.
+
+    e.g.
+    value = 5
+    [1., 2., 3., 4., 8., 9.]
+    split = 3
+
+    Args:
+        octree: A pointer to the octree where coordinates and IDs are stored
+        value: The value to split the data along
+        start: The first particle in the current node we are splitting
+        end: The last particle in the current node we are spliting
+    """
+    cdef np.int64_t index
+    cdef np.int64_t split = start
 
-    return 0
+    cdef np.float64_t * pos = octree.pposx
+    cdef np.int64_t * pidx = octree.pidx
+
+    for index in range(start, end):
+        if pos[pidx[index]] < value:
+            pidx[split], pidx[index] = pidx[index], pidx[split]
+            split+=1
+
+    return split
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+@cython.cdivision(True)
+cdef int seperate_y(Octree * octree, double value, np.int64_t start, np.int64_t end) nogil except -1:
```

Comment:
This function looks exactly like `separate_x` and `separate_z`. Shouldn't they be merged into only one?